### PR TITLE
added challenge response algo

### PR DESCRIPTION
**Pull request by snoremaster3000**
added --unlock option to do challenge/response on newer LGs

### CODE REVIEW

File: lglaf.py [hunk 1]
```diff
@@ -366,6 +367,44 @@ def command_to_payload(command):
...
+    new_key = ''
+    for x in range(32,0,-1):
+	new_key += chr(ord(old_key[x-1]) - (x % 0x0C))
+    return new_key
+
+def key_xoring(key2_t, kilo_challenge):
+    key2_t_xor = ''
+    i = 0
+    while i <= 28:
+        key2_t_xor += chr(ord(key2_t[i]) ^ ord(kilo_challenge[3]))
```
Comment (Lekensteyn) [inline lglaf.py]: For Python 3 compatibility, consider using `b''` as empty byte string and using the `struct` module. Untested, but should give an idea:

    key2_t_xor = b''
    challenge_word = struct.unpack("<I", kilo_challenge)[0]
    for i in range(0, 28 + 1, 4):
        word = struct.unpack_from("!I", key2_t, i)
        key2_t_xor += struct.pack("!I", word ^ challenge_word)
    return key2_t_xor

File: lglaf.py [hunk 2]
```diff
@@ -366,6 +367,44 @@ def command_to_payload(command):
...
+	key2_t_xor += chr(ord(key2_t[i+1]) ^ ord(kilo_challenge[2]))
+	key2_t_xor += chr(ord(key2_t[i+2]) ^ ord(kilo_challenge[1]))
+	key2_t_xor += chr(ord(key2_t[i+3]) ^ ord(kilo_challenge[0]))
+	i = i + 4
+    return key2_t_xor
+
+def do_aes_encrypt(key2_t_xor):
+    plaintext = b''
+    for k in range(0,16):
+        plaintext += chr(k)
```
Comment (Lekensteyn) [inline lglaf.py]: Try `int_as_byte` instead of `chr` in order to handle Python 2/3

File: lglaf.py [hunk 3]
```diff
@@ -366,6 +367,44 @@ def command_to_payload(command):
...
+    obj = AES.new(key2_t_xor, AES.MODE_ECB)
+    return obj.encrypt(plaintext)
+
+def do_challenge_response(comm):
+    request_kilo = make_request(b'KILO', args=[b'CENT', b'\0\0\0\0', b'\0\0\0\0', b'\0\0\0\0'])
+    kilo_header, kilo_response = comm.call(request_kilo)
+    kilo_challenge = kilo_header[8:12]
+    chalstring = ":".join("{:02x}".format(ord(k)) for k in kilo_challenge)
```
Comment (Lekensteyn) [inline lglaf.py]: This does not work with Python 3, can you change it to:

    ..format(k for k in bytearray(kilo_challenge))

similar comment for respstring below

File: lglaf.py [hunk 4]
```diff
@@ -366,6 +367,44 @@ def command_to_payload(command):
...
+    _logger.debug("Challenge: %s" %chalstring)
+    key2 = 'qndiakxxuiemdklseqid~a~niq,zjuxl' # if this doesnt work try 'lgowvqnltpvtgogwswqn~n~mtjjjqxro'
```
Comment (Lekensteyn) [inline lglaf.py]: Probably needs `key2 = b'qn...'` for Python 3 compat

**Comment by steadfasterX** [comment]
What exactly is the meaning of using --unlock? will it unlock a locked device or what?

**Comment by Lekensteyn** [comment]
@steadfasterX I think it "unlocks" some commands in the current session which would otherwise be disallowed (and this would probably not persist after a reboot).

**Comment by steadfasterX** [comment]
True when i do not use the unlock option i cant do anything in the shell! Even ls or pwd failing with an error but when I use the unlock option it works! OK the name is misleading then I think --auth would be better if that's all it does.. 
I have a lg g4 device.

**Comment by anarcat** [comment]
using this, i get a little further in rooting this LG G3 D852 phone. Unfortunately, it doesn't actually works:

```
$ sudo python lglaf.py  --unlock
LGLAF.py by Peter Wu (https://lekensteyn.nl/lglaf)
Type a shell command to execute or "exit" to leave.
# sh /data/local/tmp/lg_root.sh dummy 1 /data/local/tmp/UPDATE-SuperSU-v2.46.zip /data/local/tmp/busybox
Hello, I am LAF. Nice to meet you.
```

**Comment by Lekensteyn** [comment]
This was merged into #27, thank you for the proposal (you got acknowledged in the commit message)!

**Pull request closed without merging**